nautobot/extras/models/groups.py

```diff
@@ -431,3 +443,341 @@ def clean(self):
+    # ancestors/descendants and make this a private method?
+    def flatten_tree(self, tree, nodes=None, descendants=True, depth=1):
+        """
+        Recursively flatten a tree mapping to a list, adding a `depth` attribute to each instance in
+        the list that can be used for visualizing tree depth.
+
+        :param tree:
+            A nested dictionary tree
+        :param nodes:
+            An ordered list used to hold the flattened nodes
+        :param descendants:
+            Whether to traverse descendants or ancestors. If not set, defaults to descendants.
+        :param depth:
+            The tree traversal depth
+        """
+
+        if nodes is None:
+            nodes = []
+
+        if descendants:
+            method = "get_descendants"
+        else:
+            method = "get_ancestors"
+
+        for item in tree:
+            item.depth = depth
+            nodes.append(item)
+            branches = getattr(item, method)()
+            self.flatten_tree(branches, nodes=nodes, descendants=descendants, depth=depth + 1)
+
+        return nodes
+
+    def _ordered_filter(self, queryset, field_names, values):
+        """
+        Filters the provided `queryset` using `{field_name}__in` expressions for each field_name in the
+        list of `field_names`. The query constructed by this method explicitly orders the results in
+        the same order as the provided `values` using their list index. This is ideal for
+        maintaining ordering of topologically sorted nodes.
+
+        For example, the following would return an ordered queryset following the order in the list
+        of "pk" values:
+
+            self._ordered_filter(self.__class__.objects, ["pk"], pk_list)
+
+        :param queryset:
+            QuerySet object
+        :param field_names:
+            List of field names
+        :param values:
+            Ordered list of values corresponding values used to establish the queryset
+        """
+        if not isinstance(field_names, list):
+            raise TypeError("Field names must be a list")
+
+        case = []
+
+        # This is queryset magic to build a query that explicitly orders the items in the list of
+        # values based on their index value (idx). It's how we can get an explicitly ordered
+        # queryset that can be used for topological sort, but also supports queryset filtering.
+        for idx, value in enumerate(values):
+            when_condition = {field_names[0]: value, "then": idx}
+            case.append(models.When(**when_condition))
+
+        order_by = models.Case(*case)
+        filter_condition = {field_name + "__in": values for field_name in field_names}
+
+        return queryset.filter(**filter_condition).order_by(order_by)
+
+    def ordered_queryset_from_pks(self, pk_list):
+        """
+        Generates a queryset ordered by the provided list of primary keys.
```

Comment:
So is this basically just a way for us to reverse-engineer a QuerySet from a list? It *feels* wrong to me to do so, given the whole point of a QuerySet is to provide an efficient way to look up a list of objects from the DB, but here we have to get the whole list first and then wrap it into a QuerySet. Would be happy to discuss the use cases here to understand why this is needed/desirable.

Reply:
This is specifically a way to provide an explicitly topologically-sorted list of objects as a QuerySet. 

There's a lot that's not quite efficient here. If we were to do this correctly, we wouldn't be relying on lists at all and would instead implement a Node/Edge/NodeEdge pattern but for now this is sufficient.

In keeping with your feedback, I've changed the methods that utilize this to getters and methods, not properties. e.g. `ancestors (property)` is now `get_ancestors_queryset (method)`.